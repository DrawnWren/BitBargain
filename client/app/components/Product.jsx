import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux'; 


const mapStateToProps = state => {
  return {
    product: state.product
  };
};

const mapDispatchToProps = dispatch => {
  return {
    updateProduct: (data) => {
      dispatch({type: 'updateProduct', product: data});
    },
    clearProduct: ()=> {
      dispatch({type: 'clearProduct'});
    }
  };
};


class Product extends React.Component {
  constructor(props) {
    super(props);
  }
  componentWillMount() {
    console.log('should update the product with an ajax call here');
    fetch(`/items/${this.props.params.id}`)
      .then(res => {
	return res.json();
      })
      .then(res => {
	this.props.updateProduct(JSON.parse(res));
      })
      .catch(err => console.error(err));
  }
  componentWillUnmount() {
    this.props.clearProduct();
  }
  render() {
    return (
      <div className="container">
        <div className="card large">
          <center>
            <div className="card-image waves-effect waves-block waves-light">
              <img src={ this.props.product.images[0] } role="presentation" className="activator center" style={{width:35+'em'}}/>
            </div>
            <h2> { this.props.product.title }</h2>
          </center>
          <div className="card-reveal">
            <span className="card-title grey-text text-darken-4">{ this.props.product.title }<i className="material-icons right">close</i></span>
            <p> { this.props.product.desc } </p>
          </div>
          <div className="card-action">
            <small><p> Sold by: { this.props.product.seller } in { this.props.product.location }<br/> { this.props.product.postedAt.toString() } </p></small>
            {this.props.product.category.map((cat)=>(
              <div className="chip" key={cat}>{cat}</div>
            ))}
          </div>
        </div>
      </div>
    );
  }
}

Product.propTypes = {
  product: PropTypes.object.isRequired,
};

Product = connect(mapStateToProps, mapDispatchToProps)(Product);
export { Product };
