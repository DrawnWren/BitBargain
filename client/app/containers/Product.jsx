import React, { PropTypes } from 'react';
// import * as Default from '../components/DefaultProduct.jsx';
// import * as Buyer from '../components/BuyerProduct.jsx';
// import * as Seller from '../componentsSellerProduct.jsx';
import { connect } from 'react-redux';
import {
  mapStateToProps,
  mapDispatchToProps,
  productReducer } from '../reducers/product.reducer';
import item from '../schema';

// TODO: Get this to check current product first before loading.
// TODO: This should redirect in the event of an error.

class productContainer extends React.Component {
  componentWillMount() {
    fetch(`/items/${this.props.params.id}`)
      .then(res => res.json())
      .then(res => this.props.updateProduct(res))
      .catch(err => console.error(err));
  }
  componentWillUnmount() {
    this.props.clearProduct();
  }
  render() {
    if (this.props.user === 'buyer') {
      return (
        <Buyer
          updateProduct={this.props.updateProduct}
          product={this.props.product}
        />
      );
    } else if (this.props.user === 'seller') {
      return (
        <Seller
          updateProduct={this.props.updateProduct}
          product={this.props.product}
        />
      );
    } else {
      return (
        <Default
          updateProduct={this.props.updateProduct}
          product={this.props.product}
        />
      );
    }
  }
}

productContainer.propTypes = {
  product: item,
  params: PropTypes.shape({
    id: PropTypes.any
  }),
  updateProduct: PropTypes.func.isRequired,
  clearProduct: PropTypes.func.isRequired
};

const Product = connect(mapStateToProps, mapDispatchToProps)(productContainer);

export { Product, productReducer, productContainer };
