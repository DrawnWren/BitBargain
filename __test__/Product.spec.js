import React from 'react';


import { Product, productContainer } from '../client/app/components/Product.jsx';

import renderer from 'react-test-renderer';


// const productComponent = require('../client/app/components/Product');

it('renders correctly', () => {

    const product = { 
        image: {
        url: 'http://25.media.tumblr.com/1ed3f564cd07a5df56d845a49cc46281/tumblr_mg8k8zCiAW1qd7h1xo1_500.gif'
      },
      title: 'different Product',
      seller: 'SELLER NOME',
      desc: 'This is a desc of a product',
      category: ['cat1', 'cat2', 'cat3'],
      postedAt: '2016-09-27T02:36:02.020Z', 
      location: 'San Franpyscho',
      post: 'What'
    };
  console.log(renderer.create(<productContainer product={product} />).toJSON());
  const tree = renderer.create((<productContainer product={product} />)).toJSON();

  expect(tree).toMatchSnapshot();
});

