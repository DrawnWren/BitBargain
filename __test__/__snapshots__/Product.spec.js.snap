exports[`test renders correctly 1`] = `
<productContainer
  product={
    Object {
      "category": Array [
        "cat1",
        "cat2",
        "cat3"
      ],
      "desc": "This is a desc of a product",
      "image": Object {
        "url": "http://25.media.tumblr.com/1ed3f564cd07a5df56d845a49cc46281/tumblr_mg8k8zCiAW1qd7h1xo1_500.gif"
      },
      "location": "San Franpyscho",
      "post": "What",
      "postedAt": "2016-09-27T02:36:02.020Z",
      "seller": "SELLER NOME",
      "title": "different Product"
    }
  } />
`;
